### PR TITLE
Implement remote sync resource tagging

### LINKED ISSUE #1509: [Sync Optimization] Implement Location, CareTeam, Organization & Practitioner Resource Tagging
**Is your feature request related to a problem? Please describe.**
- Sync Optimization. This is mainly for syncing down to the server

**Describe the solution you'd like**
-  Tag all app-generated resources with the practitioner `location`, `organisation`, `careteam` & `id`. This will be achieved by the following
     - [ ] Create a method to tag the app-created data with the location resources the practitioner is assigned to. 
          - [ ] Add the following tag to the `meta.tag` attribute `{"system":"https://smartregister.org/","code":"<locationn_id>","display":"Practitioner Location"}`
     - [ ] Create a method to tag the app-created data with the organisation resources the practitioner is assigned to. 
          - [ ] Add the following tag to the `meta.tag` attribute `{"system":"https://smartregister.org/","code":"<organisation_id>","display":"Practitioner Organisation"}`
     - [ ] Create a method to tag the app-created data with the CareTeam resources the practitioner is assigned to. 
          - [ ] Add the following tag to the `meta.tag` attribute `{"system":"https://smartregister.org/","code":"<careteam_id>","display":"Practitioner CareTeam"}`
     - [ ] Create a method to tag the app-created data with the Practitioner resources the practitioner is assigned to.
          - [ ] Add the following tag to the `meta.tag` attribute `{"system":"https://smartregister.org/","code":"<Practitioner_id>","display":"Practitioner"}`

**Acceptance Criteria**
- [ ] All the app-generated should have the 4 tags present in the `meta` attribute. 

**Pre-requisite** 
- [ ] https://github.com/opensrp/fhircore/issues/857

### PR DESCRIPTION
**Pull request by FikriMilano**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #1509

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

**Review by pld — APPROVED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt [hunk 1]
```diff
@@ -16,14 +16,44 @@
...
-  LOCATION("Location"),
-  ORGANIZATION("Organization"),
-  PRACTITIONER("Practitioner")
+enum class SyncStrategy(val value: String, val tag: Coding) {
+  CARE_TEAM(
+    value = "CareTeam",
+    tag =
+      Coding().apply {
+        system = "https://smartregister.org/"
+        display = "Practitioner CareTeam"
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:32]: Is there a possibility for these codings to change? Do we want to configure them? Also can the SyncStrategy be part of `ApplicationConfiguration`? 

cc: @dubdabasoduba

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt [hunk 2]
```diff
@@ -294,3 +295,36 @@ fun Composition.retrieveCompositionSections(): List<Composition.SectionComponent
...
+fun Resource.addMandatoryTags(
+  syncStrategy: Map<String, List<String>>,
+) {
+  if (syncStrategy.containsKey(SyncStrategy.CARE_TEAM.value))
+    this.meta.addTag(
+      SyncStrategy.CARE_TEAM.tag.apply {
+        code = syncStrategy.getValue(SyncStrategy.CARE_TEAM.value).first()
+      }
+    )
+
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt:309]: This code is repeated below. I think we can refactor this to be more readable and avoid the many if statements. You can have the mandatory strategies in a list. Then iterate through the list with only one if statement needed.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt [hunk 3]
```diff
@@ -113,10 +133,37 @@ constructor(open val fhirEngine: FhirEngine, open val dispatcherProvider: Dispat
 
   suspend fun getBinary(id: String): Binary = fhirEngine.get(id)
 
-  suspend fun save(resource: Resource) {
+  suspend fun create(vararg resource: Resource): List<String> {
+    val syncStrategy = mutableMapOf<String, List<String>>()
+
+    sharedPreferencesHelper.read<List<String>>(
+        key = SharedPreferenceKey.PRACTITIONER_DETAILS_CARE_TEAM_IDS.name
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:140]: Why do we use different shared preferences keys from the sync strategy key thus enabling the code repetition? If we needed the suffix **_IDS** for the sharedPref keys that could have been done appropriately in code. I also think it does not matter what name we give them as long as they are unique.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt [hunk 4]
```diff
@@ -113,10 +133,37 @@ constructor(open val fhirEngine: FhirEngine, open val dispatcherProvider: Dispat
 
   suspend fun getBinary(id: String): Binary = fhirEngine.get(id)
 
-  suspend fun save(resource: Resource) {
+  suspend fun create(vararg resource: Resource): List<String> {
+    val syncStrategy = mutableMapOf<String, List<String>>()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:137]: Is the list a proper data structure to use? I can see `PRACTITIONER_DETAILS_USER_DETAIL` is a single object and needs to be wrapped into a list like the others.

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt [hunk 5]
```diff
@@ -113,10 +133,37 @@ constructor(open val fhirEngine: FhirEngine, open val dispatcherProvider: Dispat
...
+        syncStrategy[SyncStrategy.PRACTITIONER.value] = listOf(practitioner.id)
+      }
+
     return withContext(dispatcherProvider.io()) {
-      resource.generateMissingId()
-      fhirEngine.create(resource)
+      resource.map {
+        it.generateMissingId()
+        it.addMandatoryTags(syncStrategy)
+      }
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: I think the kotlin scope function `apply` is more appropriate here than lambda `map`
```suggestion
      resource.apply {
        generateMissingId()
        addMandatoryTags(syncStrategy)
      }
```

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt [hunk 6]
```diff
@@ -132,8 +179,7 @@ constructor(open val fhirEngine: FhirEngine, open val dispatcherProvider: Dispat
           fhirEngine.update(updateFrom(resource))
         }
       } catch (resourceNotFoundException: ResourceNotFoundException) {
-        resource.generateMissingId()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:135]: Do we no longer need to this line `resource.generateMissingId()`?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt (hunk 2 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt:309]: Sure, make sense

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 3 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:140]: Okay, I was thinking to use the SyncStrategy.value() as the key.
What do you think?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 5 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: Curious.
In this code, `resource` is a vararg.

Is it still appropriate to use `.apply` ?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 4 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:137]: Honestly, I did that because it's easier to check the mandatory tags later.

Is that the right thing?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 6 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:135]: Yea, we no longer need that.

Since create() already has that.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 1 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:32]: I think one thing from me is.

For the `system` to be more detailed.

Something like
`<Base Url>/practitioner-care-team`

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 5 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: Oh I see.

Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: How about `onEach` lambda is it possible to use here?

Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: Because I think we are updating each item here on the `varags`

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 6 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:135]: Noted.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 1 above)
Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:32]: @ellykits Yes, we can configure them. I think that will give us more flexibility if we needed to change them in future. 

@FikriMilano mmh we can do that but then that URL doesn't exist. My thinking was since these are custom codes then redirecting to the SmartRegister website makes sense.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 3 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:140]: I think the answer for this is dependent on my question whether we would want to have the `SyncStrategy` as part of `ApplicationConfiguration` @dubdabasoduba

File: android/engine/src/main/java/org/smartregister/fhircore/engine/p2p/dao/BaseP2PTransferDao.kt [hunk 7]
```diff
@@ -88,20 +83,6 @@ constructor(
...
-        fhirEngine.get(resource.resourceType, resource.logicalId).run {
-          fhirEngine.update(updateFrom(resource))
-        }
-      } catch (resourceNotFoundException: ResourceNotFoundException) {
-        resource.generateMissingId()
-        fhirEngine.create(resource)
-      }
-    }
-  }
-
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/p2p/dao/BaseP2PTransferDao.kt:104]: Why are we removing this?

**Review by dubdabasoduba — DISMISSED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/p2p/dao/BaseP2PTransferDao.kt (hunk 7 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/p2p/dao/BaseP2PTransferDao.kt:104]: We no longer need that plus it's a code duplication. 
That is replaced by defaultRepository.addOrUpdate()

[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/p2p/dao/BaseP2PTransferDao.kt:104]: We no longer need that plus it's a code duplication.
That is replaced by defaultRepository.addOrUpdate()

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 1 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:32]: Oh ok, so we will configure them and have that as it is for now.

We can easily change the `system` in the future if need be.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 5 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:165]: Yeah, sure

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 4 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:137]: Okay

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt (hunk 3 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:140]: Quoting from Ben:
> Yes, we can configure them. I think that will give us more flexibility if we needed to change them in future.

So this is confirmed to be configured.

**Comment by FikriMilano** [comment]
I had a discussion w Elly, where there is a dependency cycle happening.
So currently I'm refactoring that and also several more of his suggestions from our discussion.

cc @f-odhiambo @dubdabasoduba

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/app/ConfigService.kt [hunk 8]
```diff
@@ -103,10 +106,9 @@ interface ConfigService {
           // TODO: Does not support multi organization yet,
           // https://github.com/opensrp/fhircore/issues/1550
           ConfigurationRegistry.ORGANIZATION ->
-            paramsMap
-              ?.get(SharedPreferenceKey.PRACTITIONER_DETAILS_ORGANIZATION_IDS.name)
-              ?.firstOrNull()
-              ?.substringAfter("/")
+            mandatoryTags
+              .firstOrNull { it.display.contentEquals(SyncStrategy.ORGANIZATION.value) }
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/app/ConfigService.kt:110]: Should we ignore case here when comparing the texts?

Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/app/ConfigService.kt:110]: Let's ignore case, just to make sure

**Review by ellykits — APPROVED** [review]

**Comment by FikriMilano** [comment]
@pld @dubdabasoduba can we directly merge this? 

The main branch CI is failing, that's why this `codecov/project` coverage is significantly affected.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt [hunk 9]
```diff
@@ -22,8 +22,8 @@ package org.smartregister.fhircore.engine.sync
  * resource.
  */
 enum class SyncStrategy(val value: String) {
-  CARE_TEAM("CareTeam"),
-  LOCATION("Location"),
-  ORGANIZATION("Organization"),
+  CARE_TEAM("Practitioner CareTeam"),
+  LOCATION("Practitioner Location"),
+  ORGANIZATION("Practitioner Organization"),
```
[the remark this answers is not in the document]
Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:27]: I think the strategies should remain as they were.

**Review by dubdabasoduba — DISMISSED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:27]: Sure.
Just to clarify, we revert this as they were, but without the `system`.

Is that right?

File: android/engine/src/test/assets/configs/app/application_config.json [hunk 10]
```diff
@@ -10,7 +10,24 @@
...
+			"display": "Practitioner Location"
+		},
+		{
+			"system": "https://smartregister.org/",
+			"display": "Practitioner Organization"
+		},
+		{
+			"system": "https://smartregister.org/",
+			"display": "Practitioner"
+		}
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/test/assets/configs/app/application_config.json:29]: I see.
How about having it in local json file? 

Again just to clarify. That json file will have the Coding `system` and `display` only.

Is that right?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:27]: If we revert, doesn't that mean it's hardcoded in the code? instead from a JSON file?

File: android/engine/src/test/assets/configs/app/application_config.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/test/assets/configs/app/application_config.json:29]: When you say hardcoded pieces, do you mean a whole Coding of tag or just a single value of `system` which is only a string?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncStrategy.kt:27]: Sure.

So the value for `system` or `display` will be placed in strings.xml

Is that what you mean?

File: android/engine/src/test/assets/configs/app/application_config.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/test/assets/configs/app/application_config.json:29]: If it's a single value of a property, then yes I think it's fine to save it in strings.xml

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/login/LoginViewModel.kt [hunk 11]
```diff
@@ -120,25 +122,21 @@ constructor(
...
+    val locationIds =
+      defaultRepository.create(*locations.toTypedArray()).map { it.getSubstringBetween("/", "/") }
 
+    sharedPreferences.write(SyncStrategy.PRACTITIONER.value, practitionerDetails.userDetail)
+    sharedPreferences.write(SyncStrategy.CARETEAM.value, careTeamIds)
+    sharedPreferences.write(SyncStrategy.ORGANIZATION.value, organizationIds)
+    sharedPreferences.write(SyncStrategy.LOCATION.value, locationIds)
     sharedPreferences.write(
-      SharedPreferenceKey.PRACTITIONER_DETAILS_USER_DETAIL.name,
-      practitionerDetails.userDetail
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/ui/login/LoginViewModel.kt:129]: I think we should be saving `practitionerDetails` here instead. The ` practitionerDetails.userDetail` is a `KeycloakUserDetails` which is a property of `PractitionerDetails`. Using `KeycloakUserDetails` id, we should be using the Practitioner ID instead.

I have fixed this in a separate branch.

Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/login/LoginViewModel.kt:129]: Sure, that's ok

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package org.smartregister.fhircore.engine.util
+
+import android.app.Application
+
+object ApplicationUtil {
+  /** This Application is used to retrieve string resources outside of the Activity. */
+  lateinit var application: Application
+}
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: I do not think we need this class. A utility class just to hold reference to the `Application`.

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt (hunk 12 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: Any suggestions where to put the reference?
I need the reference to be accessible from anywhere.

Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: Wait, will this work?
```suggestion
/** This Application is used to retrieve string resources outside of the Activity. */
lateinit var application: Application
```

Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: Any reasons for not creating a data class for `SyncStrategy` instead of the limited `enum class` that is currently used?. We can still leverage the multi language support of configurations and avoid this utility function entirely

Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: I think Kotlin serializations allows for providing custom serializers which we can use to serialize the `Coding` resource in the  `SyncStrategyConfig` data class

Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: @ellykits
Because @dubdabasoduba requested for the tags system and display to be retrieved from strings.xml, it will use the app flavour to configure it.

That's what he meant by configure.

Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: @ellykits @FikriMilano Moved the discussion [here](https://github.com/opensrp/fhircore/issues/1657) to allow merging of this PR

**Comment by dubdabasoduba** [comment]
@FikriMilano Please Update the PR then ping @ellykits and I to get this merged.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt (hunk 12 above)
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/ApplicationUtil.kt:24]: This is deleted

**Review by ellykits — COMMENTED** [review]
LGTM :+1:

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/login/LoginViewModel.kt [hunk 13]
```diff
@@ -120,25 +122,21 @@ constructor(
...
+    val careTeamIds =
+      defaultRepository.create(*careTeams.toTypedArray()).map { it.getSubstringBetween("/", "/") }
+    val organizationIds =
+      defaultRepository.create(*organizations.toTypedArray()).map {
+        it.getSubstringBetween("/", "/")
+      }
+    val locationIds =
+      defaultRepository.create(*locations.toTypedArray()).map { it.getSubstringBetween("/", "/") }
 
+    sharedPreferences.write(SyncStrategy.PRACTITIONER.value, practitionerDetails.userDetail)
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/login/LoginViewModel.kt]: Okay

**Review by ellykits — APPROVED** [review]
